src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -243,13 +243,35 @@ impl<'a, 'tcx> Qualifier<'a, 'tcx, 'tcx> {
             return;
         }
 
+        if self.tcx.features().const_let {
+            let mut dest = dest;
+            let index = loop {
+                match dest {
+                    // with `const_let` active, we treat all locals equal
+                    Place::Local(index) => break *index,
+                    // projections are transparent for assignments
+                    // we qualify the entire destination at once, even if just a field would have
+                    // stricter qualification
+                    Place::Projection(proj) => dest = &proj.base,
```

Comment:
Continuing with 

> the mindset when dealing with [const qualif] should be "how do I break this?"

Projections would include `Deref`, which isn't a problem because we can't have mutable references at all inside const contexts. Still something to think about.

`Downcast` also can't happen, because we'd need `match` for that.

Maybe we should just whitelist field projections for now.

Reply:
The problem is something like this:
```rust
const FOO: &(Cell<usize>, bool) = {
    let mut a = (Cell::new(0), false);
    a.1 = true; // resets `qualif(a)` to `qualif(true)`
    &{a}
};
```
I suspect that will actually compile with this PR, although I'd like a confirmation.

We could add this as a testcase, but you might be able to come up with something simpler/more general.

Reply:
This indeed compiles successfully with this PR

Reply:
My suggestion is field (or projections in general) assignment should combine qualifications  (i.e. `|=`) instead of replacing them.

Reply:
I now reuse the existing projection checks for reading values. These checks replace the qualification with `NOT_CONST`, so there's no real use in combining I think?